File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

n, m = map(int, input().split())

graph = []

for _ in range(n):
    graph.append(list(map(int, input())))

dy = [1, 0, -1, 0]
dx = [0, 1, 0, -1]

visited = [[0] * m for _ in range(n)]

def bfs(start):
    queue = deque()
    queue.append(start)
    while queue:
        yy, xx = queue.popleft()
        visited[yy][xx] = 1
        for i in range(len(dy)):
            ny = yy + dy[i]
            nx = xx + dx[i]
            if 0 <= ny < n and 0 <= nx < m:
                if graph[ny][nx] == 1 and visited[ny][nx] == 0:
                    graph[ny][nx] += graph[yy][xx]
                    queue.append((ny, nx))

bfs((0, 0))
print(graph[n - 1][m - 1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <utility>

using namespace std;

int N, M;
string x;
vector <string> graph;
vector <vector<bool>> visited;
queue <pair<pair<int, int>, int>> q;

int dy[4] = { 1, 0, -1, 0 };
int dx[4] = { 0, 1, 0, -1 };

void BFS() {

	while (!q.empty()) {
		int y = q.front().first.first;
		int x = q.front().first.second;
		int cnt = q.front().second;
		q.pop();

		if (y == N - 1 && x == M - 1) {
			cout << cnt;
		}

		for (int k = 0; k < 4; k++) {
			int yy = y + dy[k];
			int xx = x + dx[k];

			if (yy < 0 || yy >= N || xx < 0 || xx >= M || visited[yy][xx] || graph[yy][xx] != '1')
				continue;

			visited[yy][xx] = true;
			q.push({ {yy, xx}, cnt + 1 });
		}
	}
}

int main(void) {
	cin.tie(0)->sync_with_stdio(false);

	cin >> N >> M;

	visited = vector<vector<bool>>(N, vector<bool>(M, false));
	for (int i = 0; i < N; i++) {
		cin >> x;
		graph.push_back(x);
	}

	visited[0][0] = true;
	q.push({ {0, 0}, 1 });

	BFS();

	return 0;
}